File: examples/swift/BasicUsage.swift
/**
 * Basic usage example for the Renamed Swift SDK.
 *
 * This example demonstrates:
 * - Creating a client with an API key
 * - Getting user information and credits
 * - Renaming a file using AI
 * - Handling errors
 *
 * Usage:
 *   RENAMED_API_KEY=rt_... swift run BasicUsage invoice.pdf
 */

import Foundation
import Renamed

@main
struct BasicUsage {
    static func main() async {
        // Get API key from environment
        guard let apiKey = ProcessInfo.processInfo.environment["RENAMED_API_KEY"],
              !apiKey.isEmpty else {
            fputs("Please set RENAMED_API_KEY environment variable\n", stderr)
            exit(1)
        }

        // Check command line arguments
        let args = CommandLine.arguments
        guard args.count > 1 else {
            fputs("Usage: swift run BasicUsage <file>\n", stderr)
            exit(1)
        }

        let filePath = args[1]

        do {
            // Create the client
            let client = try RenamedClient(apiKey: apiKey)

            // Get user info
            print("Fetching user info...")
            let user = try await client.getUser()
            print("User: \(user.email)")
            print("Credits: \(user.credits ?? 0)")
            print()

            // Rename a file
            print("Renaming: \(filePath)")
            let fileUrl = URL(fileURLWithPath: filePath)
            let file = try FileInput(url: fileUrl)
            let result = try await client.rename(file: file)

            print("\nResult:")
            print("  Original:  \(result.originalFilename)")
            print("  Suggested: \(result.suggestedFilename)")
            if let folder = result.folderPath, !folder.isEmpty {
                print("  Folder:    \(folder)")
            }
            if let confidence = result.confidence {
                print(String(format: "  Confidence: %.1f%%", confidence * 100))
            }

        } catch let error as RenamedError {
            handleError(error)
        } catch {
            fputs("Error: \(error.localizedDescription)\n", stderr)
            exit(1)
        }
    }

    static func handleError(_ error: RenamedError) {
        switch error {
        case .authentication(let message):
            fputs("Authentication failed: \(message)\n", stderr)
            fputs("Please check your API key\n", stderr)

        case .insufficientCredits(let message):
            fputs("Insufficient credits: \(message)\n", stderr)
            fputs("Please add more credits at https://renamed.to/dashboard\n", stderr)

        case .rateLimit(let message, let retryAfter):
            fputs("Rate limit exceeded: \(message)\n", stderr)
            if let seconds = retryAfter {
                fputs("Retry after \(seconds) seconds\n", stderr)
            }

        case .validation(let message, _):
            fputs("Validation error: \(message)\n", stderr)
            fputs("Please check your file format\n", stderr)

        default:
            fputs("Error: \(error.localizedDescription)\n", stderr)
        }
        exit(1)
    }
}

File: examples/rust/basic_usage.rs
//! Basic usage example for the Renamed Rust SDK.
//!
//! This example demonstrates:
//! - Creating a client with an API key
//! - Getting user information and credits
//! - Renaming a file using AI
//! - Handling errors
//!
//! Usage:
//!   RENAMED_API_KEY=rt_... cargo run --example basic_usage -- invoice.pdf

use renamed::{RenamedClient, RenamedError};
use std::env;
use std::process;

#[tokio::main]
async fn main() {
    // Get API key from environment
    let api_key = match env::var("RENAMED_API_KEY") {
        Ok(key) if !key.is_empty() => key,
        _ => {
            eprintln!("Please set RENAMED_API_KEY environment variable");
            process::exit(1);
        }
    };

    // Check command line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run --example basic_usage -- <file>");
        process::exit(1);
    }

    let file_path = &args[1];

    // Create the client
    let client = RenamedClient::new(&api_key);

    // Get user info
    println!("Fetching user info...");
    match client.get_user().await {
        Ok(user) => {
            println!("User: {}", user.email);
            println!("Credits: {}", user.credits.unwrap_or(0));
            println!();
        }
        Err(e) => {
            handle_error(e);
        }
    }

    // Rename a file
    println!("Renaming: {}", file_path);
    match client.rename(file_path, None).await {
        Ok(result) => {
            println!("\nResult:");
            println!("  Original:  {}", result.original_filename);
            println!("  Suggested: {}", result.suggested_filename);
            if let Some(folder) = &result.folder_path {
                if !folder.is_empty() {
                    println!("  Folder:    {}", folder);
                }
            }
            if let Some(confidence) = result.confidence {
                println!("  Confidence: {:.1}%", confidence * 100.0);
            }
        }
        Err(e) => {
            handle_error(e);
        }
    }
}

fn handle_error(error: RenamedError) {
    match error {
        RenamedError::Authentication { message, .. } => {
            eprintln!("Authentication failed: {}", message);
            eprintln!("Please check your API key");
        }
        RenamedError::InsufficientCredits { message, .. } => {
            eprintln!("Insufficient credits: {}", message);
            eprintln!("Please add more credits at https://renamed.to/dashboard");
        }
        RenamedError::RateLimit { message, retry_after, .. } => {
            eprintln!("Rate limit exceeded: {}", message);
            if let Some(seconds) = retry_after {
                eprintln!("Retry after {} seconds", seconds);
            }
        }
        RenamedError::Validation { message, .. } => {
            eprintln!("Validation error: {}", message);
            eprintln!("Please check your file format");
        }
        _ => {
            eprintln!("Error: {}", error);
        }
    }
    process::exit(1);
}
